### PR TITLE
Webgl scissor manager

### PR DESCRIPTION
**Pull request by TannerRogalsky**
A Scissor Mask for the WebGLRenderer.

API:
```javascript
displayObject.mask = new PIXI.Rectangle(0, 0, 20, 20);
```

### CODE REVIEW

**Comment by englercj** [comment]
Very cool, I'm curious what benefit this gives you over using a Graphics object since you can draw a rectangle there as well.

**Comment by TannerRogalsky** [comment]
@englercj It's just a matter for performance. It's a lot less data to transfer to the GPU if I don't need to bind a buffer and can just use send 4 floats instead. Fewer draw calls and less modified GPU state.

**Comment by englercj** [comment]
@TannerRogalsky Fair enough, I would much rather use the shapes (Polygon/Rectangle/Circle/etc) for masks rather than Graphics. The big benefit of graphics is the easy-to-use drawing APIs to create polygons without knowing vertices. But I may want to make that a v4 thing, rather than mix-and-matching in v3.

@GoodBoyDigital I'll leave it up to you whether to merge this in or not.

**Comment by TannerRogalsky** [comment]
@englercj I can see that. In general I like the API for setting masks using Graphics and I think it makes a lot of sense for setting the stencil buffer but was a little dismayed by the number of draw calls required when I just wanted to mask a simple rectangle.

Maybe it should be noted that I didn't support the CanvasRenderer with this change but that's only because I wanted to keep the PR smaller. I can't think of any technical limitation that should keep this from working with both renderers (although I haven't spent much time with the PIXI Canvas stuff). If this gets merged and there's interest, I would be happy to look into that as well.

**Comment by englercj** [comment]
Yeah you are right, and I think it is a great candidate for v4 :)

**Comment by GoodBoyDigital** [comment]
Hey guys!

Thanks for this PR @TannerRogalsky! I 100% agree that scissor masking is something we need to get into pixi as often a scissor test is way faster for simple rectangles. 

@englercj - I think the cool thing about masks extending display objects is that they can be easily animated / transformed. Using shapes would probably be a little more restrictive.

How about instead of passing in a mat.Rectangle, we keep the API the same and if a user uses a graphics object as a mask we check to see if the mask is:

1: A rectangle shape.
2: It does not have any rotation / skew.

If the above is true then we can automatically use scissor masking instead. Would be pretty neat! 
What do ya think?

**Comment by TannerRogalsky** [comment]
@GoodBoyDigital I thought about that and I think it's a great idea but I wasn't sure if it would be seen as an abuse of the Graphics object. Do you imagine just checking if `graphicsData` is length one and the resembles a rectangular polygon? It might be nice to maintain the `Rectangle` API in addition to that because it's a little more succinct to just create a rectangle directly but I can make the change to check for the Graphics case.

**Comment by TannerRogalsky** [comment]
@GoodBoyDigital Is this something like what you were thinking? 838bc81

**Comment by GoodBoyDigital** [comment]
pretty much! will aim to review this over the weekend, looks good though man! Thanks for your efforts so far @TannerRogalsky :+1:

**Comment by TannerRogalsky** [comment]
My pleasure! :+1:

File: src/core/renderers/webgl/managers/MaskManager.js [hunk 1]
```diff
@@ -52,6 +69,22 @@ MaskManager.prototype.popMask = function (target, maskData)
...
     }
+    else if (maskData.type === CONST.SHAPES.RECT) {
+        this.popScissorMask(target, maskData);
+    }
+    else if (maskData.graphicsData && maskData.graphicsData.length === 1) {
+        var graphicsData = maskData.graphicsData[0];
+        if (graphicsData.type === CONST.SHAPES.RECT && maskData.rotation === 0 &&
+            maskData.skew.x === 0 && maskData.skew.y === 0)
+        {
+            // The graphics object has no rotation or skew and only a single rectangle in it
```
Comment (englercj) [inline src/core/renderers/webgl/managers/MaskManager.js]: The check above doesn't guarantee there is only a single rectangle in the Graphics object. Just that the first one is a Rectangle.

Comment (TannerRogalsky) [inline src/core/renderers/webgl/managers/MaskManager.js]: @englercj Doesn't the `maskData.graphicsData.length === 1` ensure that there's only one shape?

**Comment by GoodBoyDigital** [comment]
hey there! Looking good @TannerRogalsky ! A  few thoughts for the PR..

- I think it would be best to check the world transform of the object, that way if the rectangle is in a rotated/skewed container it would default to regular masking. Something like the below maybe?

```
var wt = maskData.worldTransform
if( !(wt.b % CONST.PI_2) && wt.c === wt.b)
{
      // only scale and transform....
}

```

- We should also make sure to apply the translation and the scale of the worldTransform to the scissor rect.

- Im guessing we can only push one scissor rect onto the stack? As you can only have one scissor rect active at one time?

Thanks all!

**Comment by GoodBoyDigital** [comment]
New thought! 

How about - Maybe for the first mask in the stack, we could Always use a scissor rect? We could get the bounds of the mask shape, and scissor rect that. Then if the mask is not a valid rect we also use the stencil buffer. Could give us a performance boost for all masks as we only render what we know will be seen?

File: src/core/renderers/webgl/managers/MaskManager.js (hunk 1 above)
Comment (englercj) [inline src/core/renderers/webgl/managers/MaskManager.js]: DERP, missed that. Carry on! :see_no_evil:

**Comment by GoodBoyDigital** [comment]
hello! this feature has landed in v4. Thanks for the original effort @TannerRogalsky :+1:

**Pull request closed without merging**